File: 1231-replace-elements-with-greatest-element-on-right-side/replace-elements-with-greatest-element-on-right-side.py
class Solution:
    def replaceElements(self, arr: List[int]) -> List[int]:
        if len(arr) <= 0:
            return []
        
        if len(arr) == 1:
            return [-1]
        
        biggest = arr[-1]
        arr[-1] = -1
        for i in range(len(arr)-2, -1, -1):
            tmp = arr[i]
            arr[i] = biggest
            biggest = max(biggest, tmp)
        return arr

File: 1231-replace-elements-with-greatest-element-on-right-side/replace-elements-with-greatest-element-on-right-side.c
/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int* replaceElements(int* arr, int arrSize, int* returnSize) {
    *returnSize = arrSize;
    int* res = malloc(arrSize * sizeof(int));

    if (arrSize <= 0)
        return res;
    
    if (arrSize == 1)
    {
        res[0] = -1;
        return res;
    }

    int biggest = arr[arrSize - 1];
    res[arrSize - 1] = -1;

    int index = arrSize - 2;
    while (index >= 0)
    {
        int tmp = arr[index];
        res[index] = biggest;
        if (tmp > biggest)
            biggest = tmp;

        // Loop Invariant
        index--;
    }

    return res;
}